### PR TITLE
schedule: balance scheduler do not depend on average score.

### PR DESCRIPTION
**Pull request by disksing**
In some scenarios, the leader scores or region scores cannot eventually converge on the average. We have to give up using average score to determine whether to schedule.

### CODE REVIEW

File: server/schedulers/balance_test.go [hunk 1]
```diff
@@ -391,7 +341,7 @@ func (s *testBalanceRegionSchedulerSuite) TestReplicas3(c *C) {
 	cache.Remove(1)
 
 	// Store 9 has different zone with other stores but larger region score than store 1.
-	tc.addLabelsStore(9, 9, map[string]string{"zone": "z2", "rack": "r1", "host": "h1"})
+	tc.addLabelsStore(9, 20, map[string]string{"zone": "z2", "rack": "r1", "host": "h1"})
```
Comment (disksing) [inline server/schedulers/balance_test.go]: Note: The previous value here does not match the description.

File: server/schedulers/utils.go [hunk 2]
```diff
@@ -168,38 +168,15 @@ func minDuration(a, b time.Duration) time.Duration {
...
-	// In this case, either sourceSizeDiff or targetSizeDiff will be negative, then obviously return false
-	sourceSizeDiff := (sourceScore - avgScore) * source.ResourceWeight(kind)
-	targetSizeDiff := (avgScore - targetScore) * target.ResourceWeight(kind)
-
-	log.Debugf("[region %d] size diff is %v and tolerant size is %v", region.GetId(), math.Min(sourceSizeDiff, targetSizeDiff), float64(region.ApproximateSize)*tolerantRatio)
-	return math.Min(sourceSizeDiff, targetSizeDiff) >= float64(region.ApproximateSize)*tolerantRatio
+	if sourceWeight == 0 {
+		return true
+	}
+	// Make sure after move, source score is still greater than target score.
```
Comment (Connor1996) [inline server/schedulers/utils.go]: please add debug log

Comment (disksing) [reply server/schedulers/utils.go]: ok. I add debug logs when _shouldBalance_ check fails: https://github.com/pingcap/pd/pull/991/commits/2e0858ba1428389b77367f22546bafecf77e7e2d

**Comment by nolouch** [comment]
LGTM

File: server/schedulers/balance_region.go [hunk 3]
```diff
@@ -117,9 +118,13 @@ func (s *balanceRegionScheduler) transferPeer(cluster schedule.Cluster, region *
 	}
 
 	target := cluster.GetStore(newPeer.GetStoreId())
-	avgScore := cluster.GetStoresAverageScore(core.RegionKind, s.selector.GetFilters()...)
-	log.Debugf("[region %d] source store id is %v, target store id is %v, average store score is %f", region.GetId(), source.GetId(), target.GetId(), avgScore)
-	if !shouldBalance(source, target, avgScore, core.RegionKind, region, opInfluence, cluster.GetTolerantSizeRatio()) {
+	log.Debugf("[region %d] source store id is %v, target store id is %v", region.GetId(), source.GetId(), target.GetId())
+
+	sourceSize := source.RegionSize + int64(opInfluence.GetStoreInfluence(source.GetId()).RegionSize)
```
Comment (Connor1996) [inline server/schedulers/balance_region.go]: How about put these calculation into shouldBalance, and shouldBalance just need to take (source, target, cluster, opInfluence)

Comment (disksing) [reply server/schedulers/balance_region.go]: Then it will need `resourceKind` and `region` too, which has more arguments than this way.

**Pull request merged**